src/blog/oilpan-pointer-compression.md

```diff
@@ -0,0 +1,247 @@
+
+:::table-wrapper
+<!-- markdownlint-disable no-inline-html -->
+|              | upper half                               | lower half     |
+| ------------ | ---------------------------------------: | -------------: |
+| base         | <tt>U<sub>31</sub>...U<sub>1</sub>1</tt> | <tt>1...1</tt> |
+<!-- markdownlint-enable no-inline-html -->
+:::
+
+
+The decompression operation first sign extends the compressed value and then left-shifts to undo the compression operation for the sign bit. The resulting intermediate value is encoded as follows
+
+:::table-wrapper
+<!-- markdownlint-disable no-inline-html -->
+|              | upper half     | lower half                                |
+| ------------ | -------------: | ----------------------------------------: |
+| heap pointer | <tt>1...1</tt> | <tt>L<sub>31</sub>...L<sub>2</sub>00</tt> |
+| `nullptr`    | <tt>0...0</tt> | <tt>0...000</tt>                          |
+| sentinel     | <tt>0...0</tt> | <tt>0...010</tt>                          |
+<!-- markdownlint-enable no-inline-html -->
+:::
+
+Finally, the decompressed pointer is just the result of a bitwise and between this intermediate value and the base pointer.
+
+:::table-wrapper
+| C++                                                    | x64 assembly       |
+| :----------------------------------------------------- | :----------------- |
+| ```cpp                                                 | ```asm             \
+| void* Decompress(uint32_t compressed) {                | movsxd rax, edi    \
+|   uintptr_t intermediate =                             | add rax, rax       \
+|       (uintptr_t)((int32_t)compressed) << 1;           | and rax, qword ptr \
+|   return (void*)(intermediate & base);                 |     [rip + base]   \
+| }                                                      | ```                \
+| ```                                                    |                    |
+:::
+
+The resulting scheme handles cases 1.-3. transparently via a branchless asymmetric scheme.  Compression uses 3 bytes, not counting the initial register move as the call would anyways be inlined.  Decompression uses 13 bytes, counting the initial sign-extending register move.
+
+## Selected details
+
+The previous section explained the compression scheme used.  A compact compression scheme is necessary to achieve high performance.  The compression scheme from above still resulted in observable regressions in Speedometer.  The following paragraphs explain a few more tidbits needed to improve performance of Oilpan to an acceptable level.
+
+### Optimizing cage base load
+
+Technically, in C++ terms, the global base pointer can’t be a constant, because it is initialized at runtime after `main()`, whenever the embedder initializes Oilpan.  Having this global variable  mutable would inhibit the important const propagation optimization, e.g. the compiler cannot prove that a random call doesn’t modify the base and would have to load it twice:
```

Comment:
Please remove the extra space between "variable" and "mutable" in the second sentence.